<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1">Описание проекта</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2">Обзор данных</a></span></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-3">Анализ данных</a></span></li><li><span><a href="#Общие-выводы" data-toc-modified-id="Общие-выводы-4">Общие выводы</a></span></li></ul></div>

# Описание проекта

Имеются данные крупного сервиса для чтения книг по подписке.

Задача  — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Обзор данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

# Формируем sql-запрос.
query_1 = ''' SELECT *
            FROM books
        '''
books = pd.io.sql.read_sql(query_1, con = engine)


query_2 = ''' SELECT *
            FROM authors
        '''
authors = pd.io.sql.read_sql(query_2, con = engine)


query_3 = ''' SELECT *
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query_3, con = engine)


query_4 = ''' SELECT *
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query_4, con = engine)


query_5 = ''' SELECT *
            FROM ratings
        '''
reviews = pd.io.sql.read_sql(query_5, con = engine)

In [2]:
books.to_csv('books.csv', index=False)
authors.to_csv('authors.csv', index=False)
publishers.to_csv('publishers.csv', index=False)
ratings.to_csv('ratings.csv', index=False)
reviews.to_csv('reviews.csv', index=False)

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [3]:
books = pd.read_csv('books.csv')
books.info() 
books.head() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [4]:
authors = pd.read_csv('authors.csv')
authors.info() 
authors.head() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [5]:
publishers = pd.read_csv('publishers.csv')
publishers.info() 
publishers.head() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [6]:
ratings = pd.read_csv('ratings.csv')
ratings.info() 
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [7]:
reviews = pd.read_csv('reviews.csv')
reviews.info() 
reviews.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

# Анализ данных

**Посчитайте, сколько книг вышло после 1 января 2000 года**

In [9]:
books_2000 = ''' SELECT COUNT(book_id)
            FROM books
            WHERE publication_date > '2000-01-01'
        '''
pd.io.sql.read_sql(books_2000, con = engine)

count
0    819

819 книг было выпущено после 1 января 2000 г.

**Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [51]:
mean_rating = '''SELECT ratings.book_id AS book_id,
        COUNT(reviews.text) AS total_reviews,
        AVG(ratings.rating) AS mean_rating
        FROM ratings LEFT JOIN reviews ON (reviews.book_id = ratings.book_id 
            AND reviews.username = ratings.username)
        GROUP BY ratings.book_id
        ORDER BY total_reviews DESC
'''
pd.io.sql.read_sql(mean_rating, con = engine)

book_id  total_reviews  mean_rating
0        948              7     3.662500
1        695              6     4.081081
2        299              6     4.287500
3        696              6     3.830508
4        734              6     4.206897
..       ...            ...          ...
995      672              0     5.000000
996      808              0     3.000000
997      221              0     4.000000
998      387              0     4.000000
999       83              0     3.666667

[1000 rows x 3 columns]

Самое большое количество обзоров на книгу - 7. Средние оценки для всех книг находятся в диапазоне от 1,5 до 5.

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [19]:
max_publisher = ''' SELECT publisher,
                           COUNT (book_id) as book_count
                    FROM books
                    LEFT JOIN publishers ON publishers.publisher_id=books.publisher_id
                    WHERE num_pages > 50
                    GROUP BY publisher 
                    ORDER BY book_count DESC
                    LIMIT 1;
''' 
pd.io.sql.read_sql(max_publisher, con = engine)

publisher  book_count
0  Penguin Books          42

Издательство Penguin Books выпустило больше всего книг толще 50 страниц - 42 штуки.

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [27]:
author_max_ratings = '''SELECT authors.author, 
                               AVG(book_avg_ratings.avg_rating) as avg_rating
                    FROM books
                    INNER JOIN (SELECT book_id, 
                                       AVG(rating) as avg_rating
                                FROM ratings
                                GROUP BY book_id
                                HAVING COUNT(*) >= 50) AS book_avg_ratings
                    ON book_avg_ratings.book_id = books.book_id
                    INNER JOIN authors ON authors.author_id = books.author_id
                    GROUP BY authors.author
                    ORDER BY avg_rating desc
                    LIMIT 1;
'''
pd.io.sql.read_sql(author_max_ratings , con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

Джоан Роулинг оказалась явтором с самыми высокими средними оценками. 

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.**

In [44]:
users_reviews = '''SELECT ROUND(AVG(total.count)) as avg_review_count
                    FROM 
                        (SELECT sum.username, 
                                COUNT(sum.username)
                         FROM 
                            (SELECT username, 
                                    COUNT(rating_id) AS count
                             FROM ratings 
                             GROUP BY username) AS sum 
                    JOIN reviews ON reviews.username = sum.username
                    WHERE sum.count > 50
                    GROUP BY sum.username) AS total;
            
'''

pd.io.sql.read_sql(users_reviews, con = engine)

avg_review_count
0              24.0

Пользователи, поставившие больше 50 оценки, в среднем пишут 24 обзора. 

# Общие выводы

По итогам проведенного анализа были сделаны следующие выводы:

* 819 книг было выпущено после 1 января 2000 г.
* Самое большое количество обзоров на книгу - 7. Средние оценки для всех книг находятся в диапазоне от 1,5 до 5.
* Издательство Penguin Books выпустило больше всего книг толще 50 страниц - 42 штуки.
* Джоан Роулинг оказалась явтором с самыми высокими средними оценками.
* Пользователи, поставившие больше 50 оценки, в среднем пишут 24 обзора.